Dockerfile

```diff
@@ -1,54 +1,109 @@
+    git
 
-FROM alpine:3.17
+# install mimalloc for musl
+WORKDIR ${GOPATH}/src/mimalloc
+RUN set -eux &&\
+    git clone --depth 1 --branch v2.0.9 \
+        https://github.com/microsoft/mimalloc . &&\
+    mkdir -p build &&\
+    cd build &&\
+    cmake .. &&\
+    make -j$(nproc) &&\
+    make install
 
-RUN apk add wget lz4 aria2 curl jq gawk coreutils "zlib>1.2.12-r2" "libssl1.1>1.1.1q-r0" && apk update
+# download dependencies to cache as layer
+WORKDIR ${GOPATH}/src/app
+COPY go.mod go.sum ./
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/root/go/pkg/mod \
+    go mod download -x
 
-RUN addgroup terra \
-    && adduser -G terra -D -h /terra terra
+# Cosmwasm - Download correct libwasmvm version and verify checksum
+RUN set -eux &&\
+    WASMVM_VERSION=$(go list -m github.com/CosmWasm/wasmvm | cut -d ' ' -f 2) && \
+    WASMVM_DOWNLOADS="https://github.com/CosmWasm/wasmvm/releases/download/${WASMVM_VERSION}"; \
+    wget ${WASMVM_DOWNLOADS}/checksums.txt -O /tmp/checksums.txt; \
+    if [ ${BUILDPLATFORM} = "linux/amd64" ]; then \
+        WASMVM_URL="${WASMVM_DOWNLOADS}/libwasmvm_muslc.x86_64.a"; \
+    elif [ ${BUILDPLATFORM} = "linux/arm64" ]; then \
+        WASMVM_URL="${WASMVM_DOWNLOADS}/libwasmvm_muslc.aarch64.a"; \      
+    else \
+        echo "Unsupported Build Platfrom ${BUILDPLATFORM}"; \
+        exit 1; \
+    fi; \
+    wget ${WASMVM_URL} -O /lib/libwasmvm_muslc.a; \
+    CHECKSUM=`sha256sum /lib/libwasmvm_muslc.a | cut -d" " -f1`; \
+    grep ${CHECKSUM} /tmp/checksums.txt; \
+    rm /tmp/checksums.txt
 
-WORKDIR /terra
+###############################################################################
 
-COPY --from=go-builder /code/build/terrad /usr/local/bin/terrad
+FROM builder-stage-1 as builder-stage-2
 
-USER terra
+ARG GOOS=linux \
+    GOARCH=amd64
 
-# rest server
-EXPOSE 1317
-# grpc
-EXPOSE 9090
-# tendermint p2p
-EXPOSE 26656
-# tendermint rpc
-EXPOSE 26657
+ENV GOOS=$GOOS \ 
+    GOARCH=$GOARCH
 
-CMD ["/usr/local/bin/terrad", "version"]
+# Copy the remaining files
+COPY . .
+
+# Build app binary
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/root/go/pkg/mod \
+    go install \
+        -mod=readonly \
+        -tags "netgo,muslc" \
+        -ldflags " \
+            -w -s -linkmode=external -extldflags \
+            '-L/go/src/mimalloc/build -lmimalloc -Wl,-z,muldefs -static' \
+            -X github.com/cosmos/cosmos-sdk/version.Name='terrad' \
+            -X github.com/cosmos/cosmos-sdk/version.AppName='terrad' \
+            -X github.com/cosmos/cosmos-sdk/version.Version=${GIT_VERSION} \
+            -X github.com/cosmos/cosmos-sdk/version.Commit=${GIT_COMMIT} \
+            -X github.com/cosmos/cosmos-sdk/version.BuildTags='netgo,muslc' \
+        " \
+        -trimpath \
+        ./...
+
+################################################################################
+
+FROM alpine:${ALPINE_VERSION} as terra-core
+
```

Comment:
just need to copy over:
1. RUN apk add wget lz4 aria2 curl jq gawk coreutils "zlib>1.2.12-r2" "libssl1.1>1.1.1q-r0" && apk update

2. RUN addgroup terra \
    && adduser -G terra -D -h /terra terra

Reply:
these tools and user are used in many other workflow

Reply:
make sense